src/controllers/hid/hidcontroller.cpp

```diff
@@ -295,6 +239,31 @@ int HidController::close() {
     return 0;
 }
 
+bool HidController::poll() {
+    // Blocked polling: The only problem with this is that we can't close
+    // the device until the block is released, which means the controller
+    // has to send more data
+    //result = hid_read_timeout(m_pHidDevice, m_pPollData, 255, -1);
+
+    // This relieves that at the cost of higher CPU usage since we only
+    // block for a short while (500ms)
+    int result = hid_read_timeout(m_pHidDevice, m_pPollData, 255, 500);
```

Comment:
500ms is not a short while. However, I'm confused about whether this is blocking because the HIDAPI documentation says to pass a value of -1 for a blocking call:
```
		/** @brief Read an Input report from a HID device with timeout.

			Input reports are returned
			to the host through the INTERRUPT IN endpoint. The first byte will
			contain the Report number if the device uses numbered reports.

			@ingroup API
			@param device A device handle returned from hid_open().
			@param data A buffer to put the read data into.
			@param length The number of bytes to read. For devices with
				multiple reports, make sure to read an extra byte for
				the report number.
			@param milliseconds timeout in milliseconds or -1 for blocking wait.

			@returns
				This function returns the actual number of bytes read and
				-1 on error.
		*/
```
How does this fit with the 1 ms or 5 ms polling frequency set in ControllerManager?

Reply:
For what it’s worth, I copied this from the original HidReader code. It does seem like 500ms is quite a long timeout time indeed. Perhaps this can be made shorter.

I believe polling gets “paused” by Mixxx if the polling code can’t keep up, but I haven’t gone in depth into that code yet.

Reply:
Yes, I do remember writing that part. The "blocking" in the comment refers to blocking this Controller thread, not as far as HIDAPI is concerned. This call waits for up to 500ms for data from the controller instead of forever. (See above this for the truly blocked call that is commented out due to the reason stated.)

Reply:
Please correct me if I am misunderstanding, but I think this read blocking for up to 500 ms means that writes could have to wait up to 500 ms now that they are done in the same thread. Would it be better to do nonblocking reads?

Reply:
Yeah, I think the blocking read made sense in the old context with the separate HidReader thread that could wait until a new message arrived. But now that HidController is polling rather than waiting, I think a nonblocking call is more appropriate. This is what PortMidiController does. However I notice that BulkController and Hss1394Controller both work like HidController did before this PR, waiting for input rather than polling at a regular interval. Why has this been done differently for different protocols? What are the tradeoffs between regular polling with a nonblocking call versus running a while loop with blocking calls in its own thread?

Reply:
@daschuer do you have any thoughts on this? Are we going in the right direction here making HidController work like PortMidiController?

Reply:
From the HIDAPI author:

> Don't mix read/writes on multiple threads.

https://github.com/signal11/hidapi/issues/72#issuecomment-7113057

So I think we are on the right track here.

Reply:
Hm, I think you're right, a nonblocking call might be more appropriate here. I think right now it will block the entire controller thread for up to 500ms if there is no message received from the controller. My controller is very spammy so it's not a problem for me, but I'm sure that's not the case for every other controller.

Also, correct me if I'm wrong, but I think the code should maybe also keep reading in the poll function until no more messages can be received? (Should `hid_read` be wrapped in a loop until it returns 0 or -1?) What happens if the controller sends too many messages for a `poll()` call to handle? Will it start lagging behind?

Reply:
> Also, correct me if I'm wrong, but I think the code should maybe also keep reading in the poll function until no more messages can be received? (Should hid_read be wrapped in a loop until it returns 0 or -1?)

I'm not sure. The HIDAPI documentation does not make this clear. I think you'll have to experiment.

Reply:
All invocations in poll() must be non-blocking! Waiting for data to arrive is not allowed and would block other controllers. The cycle time for polling is defined by ControllerManager.

Ideally you should consume all data that is immediately available during a single invocation of poll(), i.e. read incoming data in a loop. You may limit the number of rounds for safety and to ensure that the *yet-to-be-defined* real-time constraints for the poll() task are met.